### DISCUSSION TITLE
Extending ray "payload" to multiple values per ray

### DISCUSSION THREAD

**Discussion by xntric-ngneer** [mitsuba-renderer/mitsuba3#839] [Q&A]
Hello,

I'm still a beginner at Mitsuba. For my experiments, I would like to extend the scalar value of light intensity encoded per wavelength to a tuple or an array. A bit in the same vein as how a 4-tuple encodes the polarization information. My goal is to describe more complex models of light propagation. This would also require defining adapted BSDFs, integrators, etc., to correctly process this information.
Is this possible? How would you recommend I go about this?

Thank you.

**Comment by njroussel** [comment]
Hi @xntric-ngneer 

I would recommend looking at this page of the documentation: https://mitsuba.readthedocs.io/en/stable/src/developer_guide/variants_cpp.html
The beginning, specifically, should be of great interest to you.

My recommendation would be to define a new variant, which will define a new `Spectrum` instantiation which carries arbitrarily more information than existing representations. The advantage of this approach is that it should not break any existing interfaces. However, as you mentioned, you will still need to handle whatever extra-processing you plan to do in every part of the system: BSDFs, integrators, etc...
